pipeline_dp/budget_accounting.py

```diff
@@ -100,7 +103,209 @@ def compute_budgets(self):
+    """Stores sensitivity and weight not exposed in MechanismSpec."""
+    sensitivity: float
+    weight: float
+    mechanism_spec: MechanismSpec
+
+
+class PLDBudgetAccountant:
+    """Manages the privacy budget for privacy loss distributions.
+
+    It manages the privacy budget for the pipeline using the
+    Privacy Loss Distribution (PLD) implementation from Google's
+    dp_accounting library.
+    """
+
+    def __init__(self,
+                 total_epsilon: float,
+                 total_delta: float,
+                 pld_discretization: float = 1e-4):
+        """Constructs a PLDBudgetAccountant.
+
+        Args:
+            total_epsilon: epsilon for the entire pipeline.
+            total_delta: delta for the entire pipeline.
+            pld_discretization: `value_discretization_interval` in PLD library.
+                Smaller interval results in better accuracy, but increases running time.
+
+        Raises:
+            ValueError: Arguments are missing or out of range.
+        """
+
+        _validate_epsilon_delta(total_epsilon, total_delta)
+
+        self._total_epsilon = total_epsilon
+        self._total_delta = total_delta
+        self._mechanisms = []
+        self.minimum_noise_std = None
+        self._pld_discretization = pld_discretization
+
+    def request_budget(self,
+                       noise_kind: NoiseKind,
+                       sensitivity: float = 1,
+                       weight: float = 1) -> MechanismSpec:
+        """Request a budget.
+
+        Constructs a mechanism spec based on the parameters.
+        Adds the mechanism to the pipeline for future calculation.
+
+        Args:
+            noise_kind: The kind of noise distribution for the mechanism.
+            sensitivity: The sensitivity for the mechanism.
+            weight: The weight for the mechanism.
+
+        Returns:
+            A "lazy" mechanism spec object that doesn't contain the noise
+            standard deviation until compute_budgets is called.
+        """
+        if noise_kind == NoiseKind.GAUSSIAN and self._total_delta == 0:
+            raise AssertionError(
+                "The Gaussian mechanism requires that the pipeline delta is greater than 0"
+            )
+        mechanism_spec = MechanismSpec(noise_kind=noise_kind)
+        mechanism_spec_internal = MechanismSpecInternal(
+            mechanism_spec=mechanism_spec,
+            sensitivity=sensitivity,
+            weight=weight)
+        self._mechanisms.append(mechanism_spec_internal)
+        return mechanism_spec
+
+    def compute_budgets(self):
+        """Computes the budget for the pipeline.
+
+        Composes the mechanisms and adjusts the amount of
+        noise based on given epsilon. Sets the noise for the
+        entire pipeline.
+        """
+        if not self._mechanisms:
+            return
+        if self._total_delta == 0:
+            dp_params = DifferentialPrivacyParameters(self._total_epsilon, 0)
```

Comment:
Sry, my previous comments where from assumptions that the dp_accounting library can handle that, I checked  it can't. In principle it's possible, but for now I think the easiest way is to do computation w/o the library at all for the case delta = 0.

The easiest approach is 
```
if delta == 0:
    minimum_noise_std = sum_weights/eps* sqrt(2) 
else:
     minimum_noise_std = self._find_minimum_noise_std()
<all code stays as now>
```

and then 

 **Explanation**:
1. That's because `laplace_paramer = sensitivity/eps` . No need to multiply on sensitivity here because we multiple on `sensitivity` when we compute `mechanism_noise_std` below. Sum_weights are needed in order to normalize.
2. sqrt(2) is needed in order to convert from Laplace parameter to standard deviation

I think all tests with delta=0 will require multiplication on sqrt(2).

Does it make sense?

Reply:
Sry, one more comment :)

Reply:
@dvadym No worries! Thank you for the reviews, I am learning a lot!!

1) I summed up all of the mechanism weights but I wonder, should these be included in the sum if they are equal to 1? Because we use default weight value = 1 for not-weighting. 

Sorry another question 😄 

Say we have 3 mechanisms all with weight=1, then for each mechanism we have 
`noise = sensitivity * ( sum_weights=3 / epsilon * sqrt(2) ) / mechanism.weight = 1` with `3` left over in the numerator. I think we do not want to have this constant factor 3 here.

If we have 3 mechanisms all with weight=2 then I think it works as intended to cancel out (normalize)?

`noise = sensitivity * ( sum_weights=6 / epsilon * sqrt(2) ) / mechanism.weight = 2` also with `3` left over in the numerator.

I could add a check to only add to the sum_weights if mechanism.weight > 1?

2) 👍 Makes sense to remove the constant 2 value from the variance!

Reply:
If weight = 1 everything is correct. It's basically normal/"naive" budget splitting. 

Assume that we the budget eps and we have 3 Laplace mechanisms and we'd like to split budget evenly between them (i.e. weights are the same in our terminology).

Then each mechansims would have eps/3 budget, so the Laplace parameters would be 

`sensitivy/(eps/3) = 3*sensitivity/eps`

When we compute using weight, it would be

`sensitivity*(sum_weight=3)/eps = 3*sensitivity/eps`
 
i.e. those computations are equivalent. We can consider that 3 factor as budget split :).

Reply:
Thanks for implementing! It was a very interesting code review!

Reply:
I see, it is part of the budget split! Thank you for clarifying and for the reviews!!